pkg/agent/agent_windows.go

```diff
@@ -111,6 +112,70 @@ func (i *Initializer) prepareHNSNetworkAndOVSExtension() error {
 	return util.PrepareHNSNetwork(subnetCIDR, i.nodeConfig.NodeTransportIPv4Addr, adapter, i.nodeConfig.UplinkNetConfig.Gateway, dnsServers, i.nodeConfig.UplinkNetConfig.Routes, i.ovsBridge)
 }
 
+func (i *Initializer) prepareVMNetworkAndOVSExtension() error {
+	klog.V(2).Info("Setting up VM network")
+	// Check whether VM Switch is created
+	exists, err := util.VMSwitchExists()
+	if err != nil {
+		return err
+	}
+	if exists {
+		vmSwitchIfName, err := util.GetVMSwitchInterfaceName()
+		if err != nil {
+			return err
+		}
+		klog.InfoS("VM switch teaming", "interfaceName", vmSwitchIfName)
+		if i.nodeConfig.UplinkNetConfig.Name != util.GenHostInterfaceName(vmSwitchIfName) {
+			return fmt.Errorf("unexpected teaming interface %s found", vmSwitchIfName)
+		}
+		return nil
+	}
+
+	// Get the uplink interface configuration
+	uplinkIface, _, _, err := util.GetInterfaceConfig(i.nodeConfig.UplinkNetConfig.Name)
+	if err != nil {
+		return err
+	}
+
+	var success = false
+	hostIfName := i.nodeConfig.UplinkNetConfig.Name
+	upLinkIfName := util.GenerateUplinkInterfaceName(hostIfName)
+	klog.InfoS("Using the interface", "hostIfName", hostIfName, "uplinkIfName", upLinkIfName)
+	// Rename interfaceName to interfaceName~
+	if err = util.RenameInterface(hostIfName, upLinkIfName); err != nil {
+		return err
+	}
+
+	defer func() {
+		if !success {
+			if err = util.RenameInterface(upLinkIfName, hostIfName); err != nil {
+				klog.ErrorS(err, "Failed to rename interface")
+			}
+		}
+	}()
+
+	klog.V(2).InfoS("Creating VM switch", "upLinkIfName", upLinkIfName)
+	if err = util.CreateVMSwitch(upLinkIfName); err != nil {
+		return fmt.Errorf("failed to create VM switch for interface %s: %v", upLinkIfName, err)
+	}
+
+	defer func() {
+		if !success {
+			if err = util.RemoveVMSwitch(); err != nil {
+				klog.ErrorS(err, "Failed to remove VMSwitch")
+			}
+		}
+	}()
+
+	uplinkMACStr := strings.Replace(uplinkIface.HardwareAddr.String(), ":", "", -1)
+	if err = util.RenameVMNetworkAdapter(util.LocalVMSwitch, uplinkMACStr, hostIfName, true); err != nil {
+		return err
```

Comment:
return fmt.Errorf( "failed to rename VMNetworkAdapter as %s: %v", hostIfName, err)

Reply:
Done